pkg/graph/resolvers/trb_admin_note.go

```diff
@@ -34,6 +46,124 @@ func CreateTRBAdminNote(ctx context.Context, store *storage.Store, trbRequestID
 	return createdNote, nil
 }
 
+func CreateTRBAdminNoteGeneralRequest(ctx context.Context, store *storage.Store, input model.CreateTRBAdminNoteGeneralRequestInput) (*models.TRBAdminNote, error) {
+	noteToCreate := models.TRBAdminNote{
+		TRBRequestID: input.TrbRequestID,
+		Category:     models.TRBAdminNoteCategoryGeneralRequest,
+		NoteText:     input.NoteText,
+	}
+	noteToCreate.CreatedBy = appcontext.Principal(ctx).ID()
+
+	createdNote, err := store.CreateTRBAdminNote(ctx, &noteToCreate)
+	if err != nil {
+		return nil, err
+	}
+
+	return createdNote, nil
+}
+
+func CreateTRBAdminNoteInitialRequestForm(ctx context.Context, store *storage.Store, input model.CreateTRBAdminNoteInitialRequestFormInput) (*models.TRBAdminNote, error) {
+	noteToCreate := models.TRBAdminNote{
+		TRBRequestID: input.TrbRequestID,
+		Category:     models.TRBAdminNoteCategoryInitialRequestForm,
+		NoteText:     input.NoteText,
+
+		AppliesToBasicRequestDetails: null.BoolFrom(input.AppliesToBasicRequestDetails),
+		AppliesToSubjectAreas:        null.BoolFrom(input.AppliesToSubjectAreas),
+		AppliesToAttendees:           null.BoolFrom(input.AppliesToAttendees),
+	}
+	noteToCreate.CreatedBy = appcontext.Principal(ctx).ID()
+
+	createdNote, err := store.CreateTRBAdminNote(ctx, &noteToCreate)
+	if err != nil {
+		return nil, err
+	}
+
+	return createdNote, nil
+}
+
+func CreateTRBAdminNoteSupportingDocuments(ctx context.Context, store *storage.Store, input model.CreateTRBAdminNoteSupportingDocumentsInput) (*models.TRBAdminNote, error) {
+	// it's valid for input.DocumentIDs to be empty; see note in acceptance criteria in https://jiraent.cms.gov/browse/EASI-3362
+	// don't need to check that all referenced documents belong to the same request; database checks that
+
+	noteToCreate := models.TRBAdminNote{
+		TRBRequestID: input.TrbRequestID,
+		Category:     models.TRBAdminNoteCategorySupportingDocuments,
+		NoteText:     input.NoteText,
+	}
+	noteToCreate.CreatedBy = appcontext.Principal(ctx).ID()
+
+	// ideally, we'd create the admin note and any links to documents in a single transaction, but we don't currently have that capability
+	// see Note [Database calls from resolvers aren't atomic]
+	// as it is, there's a potential bug if the documents are invalid; the record for the admin note will be created even though this returns an error
```

Comment:
I noticed this potential bug - with the current code, if you call the `createTRBAdminNoteSupportingDocuments` with invalid `documentIDs`, then this function will return an error (from the `store.CreateTRBAdminNoteTRBDocumentLinks()` call), but the record for the admin note itself will still be created, and it'll be visible to subsequent requests (without including the documents).

I tried creating the admin note <-> documents links first before creating the admin note itself, but then the foreign key constraints on the link tables prohibit creating the link records. Without an easy transaction wrapper, I see a few options:
1. Accept that this bug exists and move on - I don't _think_ it's likely to occur if the frontend code submits correct inputs.
2. Add a `DeleteTRBAdminNote()` storage method; if `store.CreateTRBAdminNoteTRBDocumentLinks()` returns an error, delete the incomplete admin note.
3. Write a new storage method for creating an admin note with document links that uses a transaction internally.

Thoughts?

Reply:
Could you query for those documents by ID first to ensure they exist before writing new records?

Reply:
@mynar7 Good thought, that'd be option 4. It'd add a database call and a bit of code, but that's not too bad.

Reply:
I'll go with option 4, looks like that's possible for both documents and recommendations.

@StevenWadeOddball That's an interesting idea, though I definitely don't think it's worth it for this ticket; it sounds like you're suggesting changing the existing method `func (s *Store) CreateTRBAdminNote(ctx context.Context, note *models.TRBAdminNote) (*models.TRBAdminNote, error)` to something like
```go
func (s *Store) CreateTRBAdminNote(
  ctx context.Context,
  note *models.TRBAdminNote,
  transactionContinuation func(*sql.Tx) error
) (*models.TRBAdminNote, error) {
  tx := s.db.Begin()
  defer tx.Rollback()

  // create admin note using tx instead of s.db

  err = transactionContinuation(tx)
  if err != nil {
    return nil, err
  }
  tx.Commit()
  return createdNote, nil  
}
```
and then the resolver would call it with something like
```go
store.CreateTRBAdminNote(ctx, &noteToCreate, func(tx *sql.Tx) error {
  // call store.CreateTRBAdminNoteTRBDocumentLinks()
  // somehow using tx instead of just store, which would require modifying that method
})
```
Is that what you have in mind?

Reply:
Yeah! That's pretty much it exactly. And if you wanted, you can have a convenience method to call the method without additional transactions. (please forgive the sudo code)
```go
func (s *Store) CreateTRBAdminNoteTX(
  ctx context.Context,
  note *models.TRBAdminNote,
  transactionContinuation func(*sql.Tx) error
) (*models.TRBAdminNote, error) {
...
}
```

```go
func (s *Store) CreateTRBAdminNote(
  ctx context.Context,
  note *models.TRBAdminNote,
) (*models.TRBAdminNote, error) {
return CreateTRBAdminNoteTX(ctx,note,func () { return nil})

}
```

Reply:
@StevenWadeOddball Gotcha, definitely interesting. I'll add this idea to my notes about making our storage code more transaction-capable.

Reply:
Returning to the original subject of this bug - I implemented a fix in https://github.com/CMSgov/easi-app/pull/2254/commits/ad14a3049720380dce538792ec875d62005e2d29 and https://github.com/CMSgov/easi-app/pull/2254/commits/1c13ceefb4008bdc236fcf26d59a36d589d84b5b.

The code for supporting documents and advice letters was very similar, so in the second of those commits, I pulled out a common function that works with anything that implements `IBaseStruct` (which every model that embeds a `BaseStruct` does). I think there are some other resolvers/validation checks that could use it, though I haven't checked (and I won't include any refactoring of them in this PR). If it's not worth having that function extracted out, I'm not committed to having it around, I can revert it. Also, if there's a better place than `pkg/models/` for it, I'm fine with that too.